### ISSUE TITLE
Wrong date format in s3 headers

### ISSUE THREAD

**Issue by cubic3d** [rclone/gofakes3#9]
#### What is the problem you are having with rclone?
If I serve via s3 rclone seems to use an invalid format for at least the `Last-Modified` header.

To test I created a directory `test` and file `file` in it. I set the file to a specific last-accessed and last-modified date using `touch -amt 202412092222 file`:
```bash
ls -la
drwxr-xr-x cubic cubic 60 B Sat Dec 21 23:07:20 2024  .
drwxr-xr-x cubic cubic 60 B Sat Dec 21 23:07:13 2024  ..
.rw-r--r-- cubic cubic  0 B Mon Dec  9 22:22:00 2024  file
```

If I now serve the file via http from the directory `test` using `rclone serve http test` I get:
```bash
curl --head localhost:8080/file

HTTP/1.1 200 OK
Accept-Ranges: bytes
Content-Length: 0
Last-Modified: Mon, 09 Dec 2024 21:22:00 GMT
Server: rclone/v1.68.2
Date: Sat, 21 Dec 2024 22:11:46 GMT
```

Now if I serve the directory `test` as a bucket with the file `file` in it using `rclone serve s3 .` I get:
```bash
curl --head localhost:8080/test/file

HTTP/1.1 200 OK
Accept-Ranges: bytes
Access-Control-Allow-Headers: Accept, Accept-Encoding, Authorization, Content-Disposition, Content-Length, Content-Type, X-Amz-Date, X-Amz-User-Agent, X-CSRF-Token, x-amz-acl, x-amz-content-sha256, x-amz-meta-filename, x-amz-meta-from, x-amz-meta-private, x-amz-meta-to, x-amz-security-token
Access-Control-Allow-Methods: POST, GET, OPTIONS, PUT, DELETE, HEAD
Access-Control-Allow-Origin: *
Access-Control-Expose-Headers: ETag
Content-Length: 0
Content-Type: application/octet-stream
Etag: "d41d8cd98f00b204e9800998ecf8427e"
Last-Modified: Mon, 9 Dec 2024 22:22:00 GMT
Server: AmazonS3
X-Amz-Id-2: RUNEMjk5NzlFQ0MyQkM2OEVDRDI5OTc5RUNDMkJDNjhFQ0QyOTk3OUVDQzJCQzY4RUNEMjk5NzlFQ0MyQkM2OA==
X-Amz-Request-Id: ECD29979ECC2BC68
Date: Sat, 21 Dec 2024 22:08:03 GMT
```

There are two things to note: The `Last-Modified` header from the s3 server has the value `Mon, 9 Dec 2024 22:22:00 GMT`. This is invalid, as the date, as far as I understand the corresponding RFC and [Mozilla](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Last-Modified), must be in a two-digit representation (same as the http server sends). It also differs from the http response by one hour. My time zone is GMT+1, it seems to use the local time and GMT inconsistently.

I came across this bug by using pgBackRest to backup my postgres cluster against s3 provided by rclone and it failed with `ERROR: [029]: invalid month 'ec '` during the debug. After looking through the debug log for the string I discovered that it refers to the cut off `Dec` month. It probably reads the date header in fixed offsets which always expect the date to be 2 digits long.

#### What is your rclone version (output from `rclone version`)
1.68.2

#### Which OS you are using and how many bits (e.g. Windows 7, 64 bit)
Linux 64 bit

#### Which cloud storage system are you using? (e.g. Google Drive)
Tested with local and SFTP backends

#### The command you were trying to run (e.g. `rclone copy /tmp remote:tmp`)
rclone serve s3 .

#### A log from the command with the `-vv` flag (e.g. output from `rclone -vv copy /tmp remote:tmp`)
Does not apply

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by ncw** [comment]
I've moved this issue here from the rclone repo

The culprit is here

https://github.com/rclone/gofakes3/blob/5af388db693f099d6889be05a2b5dffc91e6c275/gofakes3.go#L1137-L1144

That linked issue makes interesting reading.

First the format...

All the documentation for s3 - for Example [Head Object](https://docs.aws.amazon.com/AmazonS3/latest/API/API_HeadObject.html) has the time format in the form

    Last-Modified: Sun, 1 Jan 2006 12:00:00 GMT

So we'd have to say that is the de-facto standard for AWS S3.

And if I query an actual s3 server I receive the same format

    Last-Modified: Fri, 13 Sep 2024 18:00:13 GMT

So I think the format obeys the defacto S3 standard. This may not be RFC compliant, but it **is** the s3 standard.

In the s3 backend in rclone we use the http time parser to parse this time which copes just fine - that uses http.ParseTime

https://pkg.go.dev/net/http#ParseTime

> ParseTime parses a time header (such as the Date: header), trying each of the three formats allowed by HTTP/1.1: [TimeFormat](https://pkg.go.dev/net/http#TimeFormat), [time.RFC850](https://pkg.go.dev/time#RFC850), and [time.ANSIC](https://pkg.go.dev/time#ANSIC).

And it looks like AWS S3 is using the TimeFormat type

> It is like [time.RFC1123](https://pkg.go.dev/time#RFC1123) but hard-codes GMT as the time zone. The time being formatted must be in UTC for Format to generate the correct format.

So I think the actual format is correct.

Looking at the code, it makes an effort to change the time to UTC so the timezone looks correct too, however it doesn't test correct

```
mkdir /tmp/s3/bucket
mkdir -p /tmp/s3/bucket
echo hello > /tmp/s3/bucket/file.txt
TZ=GMT rclone serve s3 /tmp/s3
TZ=MST rclone serve s3 /tmp/s3
```

And comparing the last modified with the two timezones

GMT

```
$ curl --head http://127.0.0.1:8080/bucket/file.txt
...
Last-Modified: Sun, 22 Dec 2024 15:52:22 GMT
```

MST

```
$ curl --head http://127.0.0.1:8080/bucket/file.txt
...
Last-Modified: Sun, 22 Dec 2024 08:52:22 GMT
```

So something is clearly wrong here.

So in conclusion

1. Disagree with you about the time format - it is correct
2. Agree with you about the time being wrong

Maybe this is an rclone problem after all!

**Comment by cubic3d** [comment]
Thank you for moving it to the right place and spending time to dig trough it :)

I'm still very confused. Just to clarify: My main issue here is the missing padding `0` in `1 Jan 2006` vs. `01 Jan 2006`. You pointed to https://github.com/rclone/gofakes3/blob/5af388db693f099d6889be05a2b5dffc91e6c275/gofakes3.go#L1137-L1144 which explicitly defines the padding by the Go time layout string (the layout string is `Mon, 02 Jan 2006 15:04:05` and not `Mon, 2 Jan 2006 15:04:05`), but looses it if it serves the headers.

I also tried to query real Amazon S3 resources as well as Minio, Hetzner (Ceph based), B2 and R2:
Some examples
```
curl --head https://s3.amazonaws.com/arena-attachments/2015119/282dc4fc748b58803e1aed76e9e8a952.pdf

HTTP/1.1 200 OK
x-amz-id-2: cm+Ma3QJYlS4Im1NTYI/5QuWu7T3oAynXFgNUw2qArVz9RZacuvv5Qgg3nxCCIB2xntTRnEoFzk=
x-amz-request-id: 7D7XF0K267ZT0XD4
Date: Mon, 23 Dec 2024 11:14:22 GMT
Last-Modified: Sun, 08 Apr 2018 22:49:22 GMT
ETag: "4a8714103017c379f725dbedb35fa0b4"
x-amz-version-id: nYd1ZmLqMpWIxPkGoqB4OavsqzFSTtEF
Accept-Ranges: bytes
Content-Type: application/pdf
Content-Length: 478540
Server: AmazonS3
```

```
curl 'https://[redacted].eu.r2.cloudflarestorage.com/[redacted]' -v

[snip]
* Request completely sent off
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
< HTTP/1.1 200 OK
< Date: Mon, 23 Dec 2024 11:50:57 GMT
< Content-Length: 1313
< Connection: keep-alive
< Accept-Ranges: bytes
< ETag: "aa4dac536bb4642f24d76faa0455b5d0"
< Last-Modified: Fri, 06 Dec 2024 00:00:04 GMT
< Server: cloudflare
< CF-RAY: 8f68230e8d81dcb5-FRA
<
[snip]
```

All of them use the same and padded date format and never replied with a single digit. According to the linked issues on the s3 SDK the format was widened to RFC1123 at least in the v1 to allow other time zones than GMT. In the issue for v2 it's still claimed:

> In the case of the RFC1123 format, GMT time zone is the value expected.

So it's still GMT only but the SDK seems to parse RFC1123 anyway. The RFC1123 defines the date as `date = 1*2DIGIT month 2*4DIGIT` which allows for a single digit, but also for 2-digit years. I think this has been overlooked, as the intention was to allow other time zones and not HTTP/1.1 incompatible timestamps.

I think using the padded date format (as it should already be used according to the layout string in this module) would align with S3 and all other S3 compatible servers and be compatible with RFC1123, RFC822 and the fixed length requirements of HTTP's RFC2616:
>HTTP applications have historically allowed three different formats for the representation of date/time stamps:

>
>      Sun, 06 Nov 1994 08:49:37 GMT  ; RFC 822, updated by RFC 1123
>      Sunday, 06-Nov-94 08:49:37 GMT ; RFC 850, obsoleted by RFC 1036
>      Sun Nov  6 08:49:37 1994       ; ANSI C's asctime() format
>The first format is preferred as an Internet standard and represents a fixed-length subset of that defined by RFC 1123 [8] (an update to RFC 822 [[9]](https://www.w3.org/Protocols/rfc2616/rfc2616-sec17.html#bib9)).

Agree on the varying offset: That should not change as the files unix timestamp is in UTC. `TZ=GMT ls -la file` and `TZ=MST ls -la file` are different as expected since we change our interpretation. Using `TZ=MST rclone serve http .` and same with GMT, I can verify that the time in the header does not change and represents the actual time modified in UTC. So something with the time zone in this module lib must be wrong.

**Comment by ncw** [comment]
OK on further investigation, this is actually an rclone problem not a gofakes3 problem! I'm going to move this issue back to the rclone repo now, then post something for you to try.
